adagrams/game.py

```diff
@@ -1,11 +1,136 @@
+import random
+import copy
+
 def draw_letters():
-    pass
+    
+    letter_frequency = {
+    'A': 9, 
+    'B': 2, 
+    'C': 2, 
+    'D': 4, 
+    'E': 12, 
+    'F': 2, 
+    'G': 3, 
+    'H': 2, 
+    'I': 9, 
+    'J': 1, 
+    'K': 1, 
+    'L': 4, 
+    'M': 2, 
+    'N': 6, 
+    'O': 8, 
+    'P': 2, 
+    'Q': 1, 
+    'R': 6, 
+    'S': 4, 
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+    
+    letter_pool = []
+    letter_bank = []
+    #generate elements for letter_pool list in correct frequency
+    for letter, frequency in letter_frequency.items(): 
+        for _ in range(frequency):
+            letter_pool.append(letter)
+    #selecting the letters for gameplay from letter_pool list
+    while len(letter_bank) < 10:
+        letter = random.choice(letter_pool)
+        letter_pool.remove(letter)
+        letter_bank.append(letter)
```

Comment:
👍 Nice accurate modeling of how we would draw 10 tiles from a pile. By removing the tile picked each time, the probability of picking from among the remaining tile is accurately adjusted. However, keep in mind that `remove` ends up having to shift everything after the removed item forward in the list.

A big O analysis of this ends up being a little odd, since the while loop iterates 10 times (a constant) over the letter_pool (which also comes from a fixed chunk of data) so we could squint and call this O(1), but consider if this were moved into a helper:

```python
def make_letter_bank_destructive(num_letters, letter_pool):
    letter_bank = []
    while len(letter_bank) < num_letters:
        letter = random.choice(letter_pool)
        letter_pool.remove(letter)
        letter_bank.append(letter)
    return letter_bank
```

And we could call this as
```python
    make_letter_bank_destructive(10, letter_pool)
```

Now we can see that for the more general case (variable number of letter from a variable pool), this turns into something more like O(n*m) where n is the number of letters, and m is the length of the pool. If they are approximately equal, we could simplify this to O(n^2).

So which is is? O(1) or O(n^2)? It's a big difference!

Again, if the data is truly fixed, then it's O(1), but even within a certain complexity class, there are approaches that are more or less efficient. By generalizing to a parameterized view, we can get a sense of that efficiency. In the general case, this is more like a O(n^2) approach, so we might still challenge ourselves to think about other approaches (maybe approaches that don't rely on remove).

Reply:
Making a note for office hours: O(1) vs O(n^2) under the different scenarios makes sense. Would love to chat through other approaches as I am a bit stumped on how to select letters accurately without remove.